core/src/main/java/io/grpc/internal/CompositeReadableBuffer.java

```diff
@@ -64,92 +72,109 @@ public int readableBytes() {
-      }
-    }, length);
+    executeNoThrow(SKIP_OP, length, null, 0);
   }
 
+  private static final ReadOperation<byte[]> BYTE_ARRAY_OP = new ReadOperation<byte[]>() {
+    @Override
+    public int read(ReadableBuffer buffer, int length, byte[] dest, int offset) {
+      buffer.readBytes(dest, offset, length);
+      return offset + length;
+    }
+  };
+
   @Override
   public void readBytes(final byte[] dest, final int destOffset, int length) {
-    execute(new ReadOperation() {
-      int currentOffset = destOffset;
-      @Override
-      public int readInternal(ReadableBuffer buffer, int length) {
-        buffer.readBytes(dest, currentOffset, length);
-        currentOffset += length;
-        return 0;
-      }
-    }, length);
+    executeNoThrow(BYTE_ARRAY_OP, length, dest, destOffset);
   }
 
+  private static final ReadOperation<ByteBuffer> BYTE_BUF_OP = new ReadOperation<ByteBuffer>() {
+    @Override
+    public int read(ReadableBuffer buffer, int length, ByteBuffer dest, int unused) {
+      // Change the limit so that only lengthToCopy bytes are available.
+      int prevLimit = dest.limit();
+      ((Buffer) dest).limit(dest.position() + length);
+      // Write the bytes and restore the original limit.
+      buffer.readBytes(dest);
+      ((Buffer) dest).limit(prevLimit);
+      return 0;
+    }
+  };
+
   @Override
   public void readBytes(final ByteBuffer dest) {
-    execute(new ReadOperation() {
-      @Override
-      public int readInternal(ReadableBuffer buffer, int length) {
-        // Change the limit so that only lengthToCopy bytes are available.
-        int prevLimit = dest.limit();
-        ((Buffer) dest).limit(dest.position() + length);
-
-        // Write the bytes and restore the original limit.
-        buffer.readBytes(dest);
-        ((Buffer) dest).limit(prevLimit);
-        return 0;
-      }
-    }, dest.remaining());
+    executeNoThrow(BYTE_BUF_OP, dest.remaining(), dest, 0);
   }
 
-  @Override
-  public void readBytes(final OutputStream dest, int length) throws IOException {
-    ReadOperation op = new ReadOperation() {
-      @Override
-      public int readInternal(ReadableBuffer buffer, int length) throws IOException {
+  private static final ReadOperation<OutputStream> STREAM_OP = new ReadOperation<OutputStream>() {
+    @Override
+    public int read(ReadableBuffer buffer, int length, OutputStream dest, int unused)
+            throws IOException {
         buffer.readBytes(dest, length);
         return 0;
-      }
-    };
-    execute(op, length);
-
-    // If an exception occurred, throw it.
-    if (op.isError()) {
-      throw op.ex;
     }
+  };
+
+  @Override
+  public void readBytes(final OutputStream dest, int length) throws IOException {
+    execute(STREAM_OP, length, dest, 0);
   }
 
   @Override
-  public CompositeReadableBuffer readBytes(int length) {
+  public ReadableBuffer readBytes(int length) {
+    if (length <= 0) {
+      return ReadableBuffers.empty();
```

Comment:
It's unclear if this change improves or worsens performance. Now the type of the return value changes, so the JIT will not optimize the calling code as much. On the other hand, there is less indirection when `length` is present in a single buffer.